WooCommerce/Classes/Tools/InfiniteScroll/ScrollWatcher.swift

```diff
@@ -0,0 +1,43 @@
+import UIKit
+
+/// Watches for scroll content offset changes and notifies its subscribers if the scroll position is over a threshold.
+///
+final class ScrollWatcher {
+    /// Emits a stream of scroll position percentages from 0.0 (the beginning) to 1.0 (the end) if over a given threshold whenever the user scrolls a
+    /// `UIScrollView` or subclass.
+    var scrollTrigger: Observable<Double> {
+        scrollTriggerSubject
+    }
+    private let scrollTriggerSubject: PublishSubject<Double>
+    private let scrollPositionThreshold: Double
+
+    private var offsetObservation: NSKeyValueObservation?
+
+    /// - Parameter scrollPositionThreshold: the threshold of scroll position percentage (from 0.0 to 1.0) to trigger an event.
+    init(scrollPositionThreshold: Double = 0.7) {
+        self.scrollPositionThreshold = scrollPositionThreshold
+        self.scrollTriggerSubject = PublishSubject<Double>()
+    }
+
+    deinit {
+        offsetObservation?.invalidate()
+    }
+
+    func startObservingScrollPosition(tableView: UITableView) {
+        offsetObservation = tableView.observe(\UITableView.contentOffset, options: .new) { [weak self] tableView, change in
+            guard let self = self else {
+                return
+            }
+            guard let newContentOffsetY = change.newValue?.y else {
+                return
+            }
+            // When scrolling to the bottom of the list, the maximum content offset is at the top of the table view and thus we deduct the table view's
+            // height.
+            let contentHeight = tableView.contentSize.height - tableView.frame.height
```

Comment:
I suppose at some point we should also consider the `contentSize` to determine the threshold. I'm thinking if there are 100 "pages", the 30% would probably be too early to _trigger_ a page load. Something we can think about in the future. 😄

Reply:
yes the current implementation wouldn't work well with large amount of pages like you mentioned. after we integrate with the pagination tracker in a future PR, I'm thinking we could create an observable for the highest page number in the pagination tracker and `ScrollWatcher` can subscribe to get the latest highest page number `page: Int` and trigger the next page when the scroll position is higher than `(page - 1)/page` assuming each page is roughly the same size. any feedback is welcome!

Reply:
That sounds like a good plan. What we have here is good enough for now and we can always improve things later.